File: ByC/加油站.c
#include <stdio.h>

int canCompleteCircuit(int* gas, int gasSize, int* cost, int costSize) {
    int total_tank = 0;
    int curr_tank = 0;
    int start = 0;

    for (int i = 0; i < gasSize; i++) {
        int diff = gas[i] - cost[i];
        total_tank += diff;
        curr_tank += diff;
        if (curr_tank < 0) {
            start = i + 1;
            curr_tank = 0;
        }
    }

    return total_tank >= 0 ? start : -1;
}

File: Bypyhton/加油站.py
 

class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        for i in range(len(gas)):
            cost[i] = gas[i] - cost[i]
        
        begin = 0
        while begin < len(cost):
            if cost[begin] < 0:
                begin += 1
                continue

            timeOil = 0
            for j in range(begin, begin + len(cost)):
                timeOil += cost[j % len(cost)]
                
                if timeOil < 0:
                    begin = j + 1
                    break
            else:
                return begin

        return -1
    